portal-impl/src/com/liferay/portal/template/BaseTemplateResourceLoader.java

```diff
@@ -86,7 +85,8 @@ public boolean hasTemplateResource(String templateId) {
 	}
 
 	protected void init(
-		String name, Set<TemplateResourceParser> templateResourceParsers,
+		String name,
+		ServiceTrackerList<TemplateResourceParser> templateResourceParsers,
```

Comment:
@W2jx Here, if you feel it to be strange to have `ServiceTrackerList` as a parameter, you can keep `Set`, and in the sub-classes, use `ServiceTracker` + `Customizer` to populate the `private final Set<TemplateResourceParser>`;

Or you can declare this parameter to be `Iterable<TemplateResourceParser>`, because the only usage of this is to iterate on it.